### ISSUE TITLE
Rewrite/Refactor Jelly_Meta::field and Jelly_Meta::fields

### ISSUE THREAD

**Issue by loonies** [creatoro/jelly#75]
I'm not quite aware of original `Jelly_Meta::field` and `Jelly_Meta::fields` purpose when Jon wrote them, but IMO they don't provide clear API of what they should do.

If you look at the `Jelly_Meta::fields` code:

```
// $name property is not used
public function fields($field = NULL, $name = FALSE)
{
    if (func_num_args() == 0)
    {
        return $this->_fields;
    }

    if (is_array($field))
    {
        // Here we should throw exception if already initialized
        if ( ! $this->_initialized)
        {
            // Allows fields to be appended
            $this->_fields += $field;
            return $this;
        }
    }

    // We got so far, now what? This is code smell.
}
```

Take a look to the `Jelly_Meta::field`.

```
public function field($name, $type = FALSE, $options = array())
{
    // If $type is boolean, we're searching for a field
    if (is_bool($type)) // This should be: if ($type === FALSE)

    ... 
}
```

Let say I want to get a field. I could do:

```
// They all produce same result, this is bad API design
$this->meta()->field('foo');
$this->meta()->field('foo', TRUE);
$this->meta()->field('foo', FALSE);
```

**Please, discuss how should we design this API to make it better!**

**Comment by leth** [comment]
I changed some things earlier in 160416b2487311f4fb45.

It wasn't intended to address the API design though; I agree it needs some good thought and discussion.

**Comment by creatoro** [comment]
I'll try to summarize the two methods.

The upgrade part of the guide was written by Jon and it mentions the `fields()` method:

> Jelly_Meta->fields($field) no longer returns a field
> 
> Internally, this method was used heavily to find a particular field object by its name. It also resolves aliased fields, which is nice.
> 
> Since we added the field() method, it made sense to split these apart.

The current comments for the `fields()` method are clearly from earlier times as it still mentions returning a single field. The `$name` param is unnecessary as it was used to return the field's name if only a single field was queried.

So according to this the `field()` method is for single fields exclusively:
- if only the first param is set (the field name) or the second parameter is a boolean the field is returned
- if the second param is not boolean then it sets the field to the value of the second param

When getting a single field (second param is boolean) first it tries to get the field from the aliases. If this isn't successfull it caches the field (if it can be found among the fields) and returns it. If the second param is `TRUE` only the name of the field is returned (this is used by multiple methods in Jelly_Core_Model).

So, to sum it up...

`fields()` method should loose the `name` param and the phpdoc has to be updated.

`field()` method's workings are all right, but we could think the API over.

@loonies: the result is not the same:

```
$this->meta()->field('foo'); // returns the field object
$this->meta()->field('foo', TRUE); // returns only the name of the field
$this->meta()->field('foo', FALSE);  // returns the field object
```

As the complexity of the `type` param is to provide a shortcut to `$model->meta()->field('integer')->name` it could be pretty much removed. So we could use method like this:

```
public function field($name, $type = NULL, $options = array()) {}
```

Any thoughts?

**Comment by loonies** [comment]
I have some ideas, I'll create a gist as soon as I catch up some time

**Comment by marcalj** [comment]
I think it's safe to apply @creatoro fixes because doesn't break the API. When @loonies create the gist we can study it and propose the final solution.

**Comment by creatoro** [comment]
This is my proposed rewrite for the two methods: https://gist.github.com/1188311

@loonies: if you have the time, please share your thoughts on this and the other issue as well, since these are the two last issues remaining before we create the master branch for 3.1 and the develop branch for 3.2 and it would be nice to move on...

**Comment by marcalj** [comment]
@creatoro +1

**Comment by leth** [comment]
Why does `fields(...)` prevent existing fields from being overwritten during initialisation? What do we gain from this?

I only ask because I'm thinking of relaxing that in the polymorphism branch.
I still intend to have some restrictions though; the overwriting field should be the same type or a subclass of the overwritten field, and certain settings on the field will be preserved.

Are there any gains from adopting this on the main branch?

**Comment by marcalj** [comment]
Mmm, why not have one way for overwrite and another for appending? Seems logical and useful.

**Comment by creatoro** [comment]
@leth: to be honest I'm not sure and I didn't find anything in the git history that would explain this. Talking about overwriting: what are the benefits of having the restrictions you mentioned? Why not overwrite the field without checking and merging a bunch of things? I'm all for making things more simple...

Also, what are the benefits of overwriting?

**Comment by marcalj** [comment]
I prefer more flexibility, you never know...

Personally I could use for extending a model like Role or User (auth plugin). Also I use `$meta->field('foo')->bar` to change configuration, but it's a good point to have a proper way to do all cases.

**Comment by leth** [comment]
The only case I can imagine where you'd unavoidably call `fields(...)` twice is if you're using inheritance to avoid repeating your model declarations.

You might want to change fields because different models might have different validation rules.

Actually, when we're not doing polymorphism there's no real need to enforce things like subclasses, since each model has its own table so it can't really conflict with other models.

**Comment by creatoro** [comment]
@leth: rewrote the [gist](https://gist.github.com/1188311) to overwrite old fields with newly defined ones.

Edit: We could make overwriting optional tough by using a boolean value, like:

```
public function fields(array $fields = NULL, $overwrite = FALSE) {}
```

So, there you have it @marcalj, flexibility. Gist updated to include this.

**Comment by marcalj** [comment]
Ok so now will be able to add fields individually or in bulk. Also we can overwrite fields completely individually or in bulk.
Also we can fine tunning some field info using `field()` (as getter) and modifying the property.

I think is missing a way to delete a field, I'm not sure if it's present. But I'm conscient that it's very strange to do it, but in case...

So, if you want you can drop overwrite support on `fields`, because we have it on `field`, but I'm comfortable with your solution.

Thanks for your efforts.

**Comment by loonies** [comment]
https://gist.github.com/1193623

Some thoughts IMO:
- `Jelly_Meta::field` should only accept a field object and should not be responsibile for the field object creation, thus removed [Jelly_Meta L422](https://github.com/creatoro/kohana-jelly-for-Kohana-3.1/blob/3.1/develop/classes/jelly/core/meta.php#L422).
- Removed part where the field name is returned, there should be another way for getting the field name.
- We should allow the field overwriting, so if developer writes this code, it should be accepted:

```
$meta->field('foo', new Jelly_Field_String(array(....)));
$meta->field('foo', new Jelly_Field_Integer(array(....)));
```

Side question: What is this `if ( ! $this->initialized)` check all over the Jelly_Meta class? I have the feeling that all that process of `initialize`-ing and `finalize`-ing is cumbersome.

Thoughts?

**Comment by creatoro** [comment]
@loonies: the main differences of your implentation to mine are:
- not having overwriting optional, using it as the default behavior
- removing the field object creation

After thinking it over again I agree with both as I think overwriting is an expected behavior and field object creation is just bloat here.

Regarding initializiation and finalization: I think we should stick to the way Jelly works now and discuss this subject in the new issue you [opened](https://github.com/creatoro/kohana-jelly-for-Kohana-3.1/issues/78).

**Comment by loonies** [comment]
After another thought, as this is major API break, I think we should leave this as is for now (only docblock clean up and removing $name param etc..), and introduce this in the next major release.

**Comment by loonies** [comment]
Looks like I spoke too soon, after reading [this comment](https://github.com/creatoro/kohana-jelly-for-Kohana-3.1/issues/78#issuecomment-1996484), we can brake API now :)

**Comment by creatoro** [comment]
@loonies: I think having this done for 1.0 would be a perfect timing as that would be the first stable release. We're on a development branch now, major changes are possible and expected.

Edit: ah, your were fast :).

**Comment by leth** [comment]
Looks good, the simplicity is definitely better here :)

**Comment by creatoro** [comment]
Thanks guys!

@loonies: I've seen the new version of your gist. I thought about it, but I think reusing the `field()` method in `fields()` is more elegant. Also if you require the field to be an instance of `Jelly_Field`, then you can't set aliases.

**Comment by loonies** [comment]
Can you give an use case example of setting alias?

**Comment by creatoro** [comment]
Check `Model_Test_Post`:

```
    // Set fields
    $meta->fields(array(
        'id'              => Jelly::field('primary'),
        'slug'            => Jelly::field('slug', array(
            'unique' => TRUE
        )),

        // Alias columns, for testing
        '_id'             => 'id',
        '_slug'           => 'slug',
    ));
```

**Comment by loonies** [comment]
Where is this useful and where should we use this (use case)? I prefer clear API over confusing features (maybe I'm the only one confused :))

**Comment by creatoro** [comment]
Well, personally I don't use field aliasing, so I wouldn't mind getting rid of it if nobody else uses it. Anybody? :)

Edit: after going through the code it seems that it's a feature that is deeply built into Jelly and removing it is not an easy task. I think this should be investigated further and we should skip touching it this time.

**Comment by marcalj** [comment]
Confusion here too! :P
Good work!
